### PR TITLE
lzma: set a memory limit to liblzma

### PR DESCRIPTION
**Pull request by catenacyber**
Found by oss-fuzz
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=17044

### CODE REVIEW

File: htp/htp_decompressors.c [hunk 1]
```diff
@@ -295,12 +295,18 @@ static htp_status_t htp_gzip_decompressor_decompress(htp_decompressor_gzip_t *dr
...
+                        rc = Z_STREAM_END;
+                        break;
+                    case SZ_OK:
+                        rc = Z_OK;
+                        break;
+                    case SZ_ERROR_MEM:
+                        htp_log(d->tx->connp, HTP_LOG_MARK, HTP_LOG_WARNING, 0, "LZMA decompressor: memory limit reached");
+                        // no break
+                    default:
+                        rc = Z_DATA_ERROR;
```
Comment (victorjulien) [inline htp/htp_decompressors.c:309]: can we get a more specific error to log here? I think Z_DATA_ERROR gets us a generic error.

**Review by victorjulien — CHANGES_REQUESTED** [review]
I will rerun QA on the updated branch. It will need a bunch of updates:
1. remove the liblzma commits
2. add lzma-sdk in an isolated commit changing nothing in it. This should include the autotools integration.
3. commit to switch from liblzma to sdk
4. our changes to the sdk

I want to keep things very clearly separated as we'll likely have to reimplement some of these things if we update to a newer sdk version in the future.

File: htp/htp_decompressors.c [hunk 2]
```diff
@@ -263,30 +267,48 @@ static htp_status_t htp_gzip_decompressor_decompress(htp_decompressor_gzip_t *dr
...
+                switch (status) {
+                    case LZMA_STATUS_FINISHED_WITH_MARK:
+                        rc = Z_STREAM_END;
+                        break;
+                    case SZ_OK:
+                        rc = Z_OK;
+                        break;
+                    case SZ_ERROR_MEM:
+                        htp_log(d->tx->connp, HTP_LOG_MARK, HTP_LOG_WARNING, 0, "LZMA decompressor: memory limit reached");
+                        // no break
```
Comment (victorjulien) [inline htp/htp_decompressors.c:307]:
```
htp_decompressors.c: In function ‘htp_gzip_decompressor_decompress’:
htp_decompressors.c:306:25: error: this statement may fall through [-Werror=implicit-fallthrough=]
                         htp_log(d->tx->connp, HTP_LOG_MARK, HTP_LOG_WARNING, 0, "LZMA decompressor: memory limit reached");
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
htp_decompressors.c:308:21: note: here
                     default:
                     ^~~~~~~
```

Comment (victorjulien) [reply htp/htp_decompressors.c:307]: To fix:
```diff
diff --git a/htp/htp_decompressors.c b/htp/htp_decompressors.c
index 2896118..e48bb1c 100644
--- a/htp/htp_decompressors.c
+++ b/htp/htp_decompressors.c
@@ -304,7 +304,7 @@ restart:
                         break;
                     case SZ_ERROR_MEM:
                         htp_log(d->tx->connp, HTP_LOG_MARK, HTP_LOG_WARNING, 0, "LZMA decompressor: memory limit reached");
-                        // no break
+                        // fall through
                     default:
                         rc = Z_DATA_ERROR;
                 }
```

Comment (catenacyber) [reply htp/htp_decompressors.c:307]: ok

File: htp/htp_decompressors.c (hunk 1 above)
Comment (catenacyber) [reply htp/htp_decompressors.c:309]: That could be `Z_MEM_ERROR`

But we rather want `Z_DATA_ERROR` as it is used further to stop further decompression and call the callback with the last decompressed data

Or that could be `Z_STREAM_END`

**Comment by victorjulien** [comment]
When not bundling libhtp, things go wrong. Configure says it can't use htp because:
```
configure:21792: checking htp/htp.h usability
configure:21792: clang-8 -c -O0 -ggdb3 -Werror -Wchar-subscripts -fno-strict-aliasing -fstack-protector-all -fsanitize=address -fno-omit-frame-pointer -Wshadow -march=native -I/opt/htp-0.5.x/include/   -I/usr/src/netmap -I/opt/htp-0.5.x/include -I/opt/htp-0.5.x/lib/htp/include conftest.c >&5
In file included from conftest.c:165:
In file included from /opt/htp-0.5.x/include/htp/htp.h:55:
/opt/htp-0.5.x/include/htp/htp_decompressors.h:47:10: fatal error: 'lzma/LzmaDec.h' file not found
#include "lzma/LzmaDec.h"
         ^~~~~~~~~~~~~~~~
1 error generated.

```

**Comment by victorjulien** [comment]
This works but makes the htp_decompressors.h private. Not a problem for Suricata, but maybe others would be affected?
```diff
diff --git a/htp/htp.h b/htp/htp.h
index 36209ad..f6c3c7a 100644
--- a/htp/htp.h
+++ b/htp/htp.h
@@ -52,7 +52,6 @@ extern "C" {
 #include "htp_base64.h"
 #include "htp_config.h"
 #include "htp_connection_parser.h"
-#include "htp_decompressors.h"
 #include "htp_hooks.h"
 #include "htp_list.h"
 #include "htp_multipart.h"
diff --git a/htp/htp_private.h b/htp/htp_private.h
index 80258cb..dc82ce0 100644
--- a/htp/htp_private.h
+++ b/htp/htp_private.h
@@ -62,6 +62,7 @@ extern "C" {
 
 #include "htp_config_auto_gen.h"
 #include "htp.h"
+#include "htp_decompressors.h"
 #include "htp_config_private.h"
 #include "htp_connection_parser_private.h"
 #include "htp_connection_private.h"

```

**Comment by catenacyber** [comment]
I will try to add
```
library_includedir = $(includedir)/$(GENERIC_LIBRARY_NAME)/lzma
library_include_HEADERS = $(h_sources)
```
to Makefile.am from lzma

Another PR with reordered commits is coming

**Pull request closed without merging**